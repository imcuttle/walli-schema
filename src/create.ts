import type { Verifiable } from 'walli'
import { sync as visit } from '@moyuyc/visit-tree'
import * as mapObj from 'map-obj'
import * as clone from 'lodash.clonedeep'

export type Rule<T extends Verifiable = Verifiable> = ((rule: any, opts: any) => T) | ((...args: any[]) => T) | T

export type ToWalliOptions = {
  throwError?: boolean
  silent?: boolean
}

function isPlainObject(value) {
  if (Object.prototype.toString.call(value) !== '[object Object]') {
    return false
  }

  const prototype = Object.getPrototypeOf(value)
  return prototype === null || prototype === Object.prototype
}

export type WalliSet<T extends Verifiable = Verifiable> = {
  [type: string]: Rule<T>
}

export type WalliSchema<Keys extends string = string> = {
  $type: Keys
  rule?: (WalliSchema<Keys> & { $type?: Keys }) | any
  options?: any
  required?: boolean
  optional?: boolean
}

export function createSchemaToWalli<T extends WalliSet | any>(walliSet: T) {
  return function schemaToWalli<R extends Verifiable = Verifiable>(
    // @ts-ignore
    schema: WalliSchema<keyof T>,
    opts?: ToWalliOptions
  ): R {
    let rootVerifiable: any
    const { silent = false, throwError } = opts || {}
    if (typeof schema !== 'object' || !schema?.$type) {
      return schema as any
    }

    const cloned = clone(schema)
    visit(
      cloned,
      () => {},
      (schema: WalliSchema, ctx) => {
        let { $type, rule, options, required, optional } = schema
        if (!$type && !!ctx.parent) {
          rootVerifiable = schema
          return
        }

        const walliInstance: Rule = walliSet[$type] || walliSet[`${$type}_`]
        if (!walliInstance) {
          const msg = `$type: ${JSON.stringify($type)}, ${JSON.stringify($type + '_')} is not found.`
          if (!throwError && !silent) {
            console.error(`[walli-schema] ${msg}`)
          }
          if (throwError) {
            throw new Error(msg)
          }
          rootVerifiable = schema
          return
        }

        if (typeof walliInstance === 'function') {
          const mapRule = (rule) => {
            if (Array.isArray(rule)) {
              return rule.map((r) => mapRule(r))
            }
            if (!isPlainObject(rule)) {
              return rule
            }

            if (rule.$type) {
              return schemaToWalli(rule, opts)
            }

            return mapObj(
              rule,
              // @ts-ignore
              (key: any, sourceValue: any) => {
                return [key, mapRule(sourceValue)]
              },
              { deep: false }
            )
          }
          rootVerifiable = walliInstance(mapRule(rule), options)
          if (required) {
            rootVerifiable = rootVerifiable.required
          } else if (optional) {
            rootVerifiable = rootVerifiable.optional
          }
          ctx.replace(rootVerifiable)
          return
        }
        rootVerifiable = walliInstance
        if (required) {
          rootVerifiable = rootVerifiable.required
        } else if (optional) {
          rootVerifiable = rootVerifiable.optional
        }
        ctx.replace(rootVerifiable)
      },
      {
        path: 'rule'
      }
    )

    return rootVerifiable as R
  }
}
