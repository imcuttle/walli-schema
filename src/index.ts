/**
 * walli schema to walli
 * @author imcuttle
 */
import * as w from 'walli'
import * as clone from 'lodash.clonedeep'
import { sync as visit } from '@moyuyc/visit-tree'
import { Verifiable } from 'walli'

type Rule<T extends Verifiable = Verifiable> = ((rule: any, opts: any) => T) | ((...args: any[]) => T) | T

export type WalliSet<T extends Verifiable = Verifiable> = {
  [type: string]: Rule<T>
}

export type WalliSchema<Keys extends string = string> = {
  $type: Keys
  rule?: (WalliSchema<Keys> & { $type?: Keys }) | any
  options?: any
  required?: boolean
  optional?: boolean
}

// type Options<T> = {
//   defaultType?: keyof T
// }

type ToWalliOptions = {
  throwError?: boolean
  silent?: boolean
}

export function createSchemaToWalli<T extends WalliSet | any>(walliSet: T) {
  return function schemaToWalli<R extends Verifiable = Verifiable>(
    // @ts-ignore
    schema: WalliSchema<keyof T>,
    opts?: ToWalliOptions
  ) {
    let rootVerifiable: any
    const { silent = false, throwError } = opts || {}

    const cloned = clone(schema)
    visit(
      cloned,
      () => {},
      (schema: WalliSchema, ctx) => {
        let { $type, rule, options, required, optional } = schema
        if (!$type && !!ctx.parent) {
          rootVerifiable = schema
          return
        }

        const walliInstance: Rule = walliSet[$type] || walliSet[`${$type}_`]
        if (!walliInstance) {
          const msg = `$type: ${JSON.stringify($type)}, ${JSON.stringify($type + '_')} is not found.`
          if (!throwError && !silent) {
            console.error(`[walli-schema] ${msg}`)
          }
          if (throwError) {
            throw new Error(msg)
          }
          rootVerifiable = schema
          return
        }

        if (typeof walliInstance === 'function') {
          rootVerifiable = walliInstance(rule, options)
          if (required) {
            rootVerifiable = rootVerifiable.required
          } else if (optional) {
            rootVerifiable = rootVerifiable.optional
          }
          ctx.replace(rootVerifiable)
          return
        }
        rootVerifiable = walliInstance
        if (required) {
          rootVerifiable = rootVerifiable.required
        } else if (optional) {
          rootVerifiable = rootVerifiable.optional
        }
        ctx.replace(rootVerifiable)
      },
      {
        path: 'rule'
      }
    )

    return rootVerifiable as R
  }
}

export const schemaToWalli = createSchemaToWalli(w)
