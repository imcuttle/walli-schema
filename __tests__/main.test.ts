/**
 * @file main
 * @author imcuttle
 * @date 2018/4/4
 */
import { schemaToWalli } from '../src'

import { sync as visit } from '@moyuyc/visit-tree'

const getRules = (v) => {
  let nodes: any[] = []
  visit(
    v,
    (node) => {
      nodes.push(node)
    },
    { path: 'rule' }
  )
  return nodes
}

describe('walliSchema', function () {
  it('should spec', function () {
    expect(
      schemaToWalli({
        $type: 'string'
      })
    ).toMatchSnapshot()

    expect(
      schemaToWalli({
        $type: 'leq',
        rule: {
          abc: 'test'
        }
      }).ok({
        abc: 'test',
        x: 1
      })
    ).toBeTruthy()

    const w = schemaToWalli({
      $type: 'leq',
      rule: {
        data: {
          $type: 'leq',
          rule: {
            redirect: {
              $type: 'trueLike'
            }
          }
        }
      }
    })
    expect(
      w.ok({
        data: {
          redirect: null,
          x: {}
        }
      })
    ).toBeFalsy()

    expect(
      w.ok({
        data: {
          redirect: {
            need_redirect: {}
          }
        }
      })
    ).toBeTruthy()
  })

  it('should deep', function () {
    const verifiable = schemaToWalli({
      $type: 'leq',
      rule: {
        $type: 'every',
        rule: [
          {
            $type: 'leq',
            rule: {
              a: 1
            }
          },
          {
            $type: 'oneOf',
            rule: [
              {
                $type: 'leq',
                rule: {
                  b: 2,
                  c: 3
                }
              },
              {
                $type: 'leq',
                rule: {
                  b: 3,
                  c: 2
                }
              }
            ]
          }
        ]
      }
    })

    expect(
      verifiable.ok({
        a: 1
      })
    ).toBeFalsy()

    expect(
      verifiable.ok({
        a: 1,
        b: 2
      })
    ).toBeFalsy()
    //
    expect(
      verifiable.ok({
        a: 1,
        b: 2,
        c: 3
      })
    ).toBeTruthy()

    expect(
      verifiable.ok({
        a: 1,
        b: 3,
        c: 2
      })
    ).toBeTruthy()
  })

  it('should deep passed', function () {
    const verifiable = schemaToWalli({
      $type: 'leq',
      rule: {
        $type: 'every',
        rule: [
          {
            $type: 'leq',
            rule: {
              a: {
                x: 1,
                y: 2
              },
              d: [
                {
                  x: 1,
                  y: 2
                },
                {
                  $type: 'eq',
                  rule: {
                    x: 1,
                    y: 2
                  }
                }
              ]
            }
          },
          {
            $type: 'oneOf',
            rule: [
              {
                $type: 'leq',
                rule: {
                  b: 2,
                  c: 3
                }
              },
              {
                $type: 'leq',
                rule: {
                  b: 3,
                  c: 2
                }
              }
            ]
          }
        ]
      }
    })

    expect(getRules(verifiable)).toMatchSnapshot()
  })
})
