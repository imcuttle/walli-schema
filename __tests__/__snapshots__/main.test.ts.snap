// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`walliSchema should deep passed 1`] = `
Array [
  LooseEqual {
    "_eq": [Function],
    "isRequired": true,
    "msg": null,
    "options": Object {
      "loose": true,
    },
    "rule": Every {
      "isRequired": true,
      "msg": null,
      "options": Object {},
      "rule": Array [
        LooseEqual {
          "_eq": [Function],
          "isRequired": true,
          "msg": null,
          "options": Object {
            "loose": true,
          },
          "rule": Object {
            "a": Object {
              "x": 1,
              "y": 2,
            },
            "d": Array [
              Object {
                "x": 1,
                "y": 2,
              },
              Equal {
                "_eq": [Function],
                "isRequired": true,
                "msg": null,
                "options": Object {
                  "loose": false,
                },
                "rule": Object {
                  "x": 1,
                  "y": 2,
                },
              },
            ],
          },
        },
        OneOf {
          "isRequired": true,
          "msg": null,
          "options": Object {},
          "rule": Array [
            LooseEqual {
              "_eq": [Function],
              "isRequired": true,
              "msg": null,
              "options": Object {
                "loose": true,
              },
              "rule": Object {
                "b": 2,
                "c": 3,
              },
            },
            LooseEqual {
              "_eq": [Function],
              "isRequired": true,
              "msg": null,
              "options": Object {
                "loose": true,
              },
              "rule": Object {
                "b": 3,
                "c": 2,
              },
            },
          ],
        },
      ],
    },
  },
  Every {
    "isRequired": true,
    "msg": null,
    "options": Object {},
    "rule": Array [
      LooseEqual {
        "_eq": [Function],
        "isRequired": true,
        "msg": null,
        "options": Object {
          "loose": true,
        },
        "rule": Object {
          "a": Object {
            "x": 1,
            "y": 2,
          },
          "d": Array [
            Object {
              "x": 1,
              "y": 2,
            },
            Equal {
              "_eq": [Function],
              "isRequired": true,
              "msg": null,
              "options": Object {
                "loose": false,
              },
              "rule": Object {
                "x": 1,
                "y": 2,
              },
            },
          ],
        },
      },
      OneOf {
        "isRequired": true,
        "msg": null,
        "options": Object {},
        "rule": Array [
          LooseEqual {
            "_eq": [Function],
            "isRequired": true,
            "msg": null,
            "options": Object {
              "loose": true,
            },
            "rule": Object {
              "b": 2,
              "c": 3,
            },
          },
          LooseEqual {
            "_eq": [Function],
            "isRequired": true,
            "msg": null,
            "options": Object {
              "loose": true,
            },
            "rule": Object {
              "b": 3,
              "c": 2,
            },
          },
        ],
      },
    ],
  },
  LooseEqual {
    "_eq": [Function],
    "isRequired": true,
    "msg": null,
    "options": Object {
      "loose": true,
    },
    "rule": Object {
      "a": Object {
        "x": 1,
        "y": 2,
      },
      "d": Array [
        Object {
          "x": 1,
          "y": 2,
        },
        Equal {
          "_eq": [Function],
          "isRequired": true,
          "msg": null,
          "options": Object {
            "loose": false,
          },
          "rule": Object {
            "x": 1,
            "y": 2,
          },
        },
      ],
    },
  },
  Object {
    "a": Object {
      "x": 1,
      "y": 2,
    },
    "d": Array [
      Object {
        "x": 1,
        "y": 2,
      },
      Equal {
        "_eq": [Function],
        "isRequired": true,
        "msg": null,
        "options": Object {
          "loose": false,
        },
        "rule": Object {
          "x": 1,
          "y": 2,
        },
      },
    ],
  },
  OneOf {
    "isRequired": true,
    "msg": null,
    "options": Object {},
    "rule": Array [
      LooseEqual {
        "_eq": [Function],
        "isRequired": true,
        "msg": null,
        "options": Object {
          "loose": true,
        },
        "rule": Object {
          "b": 2,
          "c": 3,
        },
      },
      LooseEqual {
        "_eq": [Function],
        "isRequired": true,
        "msg": null,
        "options": Object {
          "loose": true,
        },
        "rule": Object {
          "b": 3,
          "c": 2,
        },
      },
    ],
  },
  LooseEqual {
    "_eq": [Function],
    "isRequired": true,
    "msg": null,
    "options": Object {
      "loose": true,
    },
    "rule": Object {
      "b": 2,
      "c": 3,
    },
  },
  Object {
    "b": 2,
    "c": 3,
  },
  LooseEqual {
    "_eq": [Function],
    "isRequired": true,
    "msg": null,
    "options": Object {
      "loose": true,
    },
    "rule": Object {
      "b": 3,
      "c": 2,
    },
  },
  Object {
    "b": 3,
    "c": 2,
  },
]
`;

exports[`walliSchema should spec 1`] = `
String {
  "isRequired": true,
  "msg": null,
  "options": Object {
    "type": [Function],
    "typeName": "string",
  },
  "rule": [Function],
  "toString": [Function],
}
`;
